De-duplicate "generate a nondet int in a given range" logic

antlechner opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->
This PR addresses code duplication between `nondet.cpp` and `java_object_factory.cpp`. The latter had a utility function `gen_nondet_int_init` which can be replaced with `generate_nondet_int` from `nondet.h` after a few small changes.
No functional change is made in this PR, only refactoring.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- n/a Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- n/a My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

antlechner committed:
Move nondet documentation to header file

Following the new coding standard

antlechner committed:
Split generate_nondet_int into two functions

We will be able to use the exprt version in other parts of the code.

antlechner committed:
Parameterize basename_prefix

This will make the functions more re-usable in parts of the code that
use different prefixes.

antlechner committed:
Use allocate_objects to manage new symbol creation

This simplifies the code in generate_nondet_init and reduces code
duplication.

antlechner committed:
Remove duplicate helper function in object factory

Calls to java_object_factoryt::gen_nondet_int_init can be replaced with
equivalent calls to generate_nondet_int from nondet.h.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 1433c1c20f92e25f15a034f12183520a122912ca).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105338866
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

romainbrenguier commented on src/util/nondet.cpp:
```diff
@@ -93,15 +82,16 @@ code_blockt generate_nondet_switch(
   code_blockt result_block;
 
+  allocate_objectst allocate_objects(
+    mode, source_location, name_prefix, symbol_table);
```
:pick: prefer braces

antlechner replied:
Done. I keep forgetting about this new convention...

romainbrenguier replied:
It's not really new, just that it wasn't really applied, then someone pointed out this guideline to me on one of my PR, so I have been annoying everyone with this since then. :smiling_imp:

tautschnig commented on src/util/nondet.h:
```diff
@@ -13,6 +13,25 @@ Author: Diffblue Ltd.
 #include <util/symbol_table.h>
 
+/// Same as \ref generate_nondet_int(
+///   const mp_integer &min_value,
```
Does this multi-line `\ref` actually render correctly in Doxygen?

antlechner replied:
I tried it locally and it links to the right function, though only the `generate_nondet_init` part turns into a link. Some of the parameter types (like `source_locationt`) also appear as links to the Doxygen for those types.
If there is a better / more standard way of linking between overloaded functions, please let me know.

tautschnig commented on src/util/nondet.cpp:
```diff
@@ -7,6 +7,7 @@ Author: Diffblue Ltd.
 \*******************************************************************/
 
 #include "nondet.h"
+#include "allocate_objects.h"
```
Nit pick:
````c++
#include "nondet.h"

#include "allocate_objects.h"
#include "arith_tools.h"
#include "c_types.h"
#include "fresh_symbol.h"
#include "symbol.h"
````

antlechner replied:
Fixed the linebreaks in the same commit and removed the "util" parts in a new commit (same as the one for the forward declarations below).

romainbrenguier commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1172,12 +1118,13 @@ void java_object_factoryt::allocate_nondet_length_array(
     max_length_expr,
-    location);
+    "nondet_array_length",
+    loc,
```
:question: why is `location` replaced with `loc`?

antlechner replied:
Didn't even notice that as CLion suggested `loc` here. `loc` is a field of the object factory and `location` is the parameter to the function. From what I can see these two values are always the same so one of them is unnecessary. For now I changed it back to `location`, and I'll tidy up the `loc`/`location` duplication in a follow-up PR. :tangerine:

tautschnig commented on src/util/nondet.h:
```diff
@@ -9,6 +9,7 @@ Author: Diffblue Ltd.
+#include <util/allocate_objects.h>
 #include <util/std_code.h>
 #include <util/std_expr.h>
 #include <util/symbol_table.h>
```
I guess it's the first time I'm properly looking at this: the only includes we really need here are `#include "std_code.h"` and `#include "std_expr.h"`. Everything else can (and IMHO: should be) done via forward declarations.

antlechner replied:
Done in a new commit.

romainbrenguier reviewed (approved):
Looks great

tautschnig commented on jbmc/src/java_bytecode/java_object_factory.cpp:
```diff
@@ -1476,12 +1423,13 @@ bool java_object_factoryt::gen_nondet_enum_init(
     minus_exprt(length_expr, from_integer(1, java_int_type())),
-    location);
+    "enum_index_init",
+    loc,
```
Just like @romainbrenguier I'm surprised this even compiles? As it does, is the `location` parameter for this method unnecessary?

antlechner replied:
See my comment marked :tangerine: above.
The way the object factory passes variables around is currently quite messy in general, it uses a mix of
- class fields (even though there isn't a publicly visible class here; from what I can see the class is only used to store fields)
- many parameters to most functions
- a poorly named `java_object_factory_parameterst` which groups further arguments into a struct.

The three common ways of moving lots of variables around are all combined in one place here... I would like to generally tidy this up a bit, probably moving away from lots of function parameters and fields to bundling more things in (a) struct(s). Let me know what you think.

antlechner committed:
Tidy up includes

Specifying "util" is not necessary as nondet.cpp is itself in the util
folder.
Forward declarations can be used in nondet.h.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: fcb2ec6482d589e0755b7ffddea5ae0bd1aff36b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105427335
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: f2c6f83a5be0a4055758418e45c4682c9e43e8fb).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105427499
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

kroening commented on src/util/nondet.cpp:
```diff
@@ -60,30 +38,35 @@ symbol_exprt generate_nondet_int(
+  allocate_objectst &allocate_objects,
+  code_blockt &instructions)
+{
+  PRECONDITION(min_value < max_value);
```
`<=` is fine

antlechner replied:
Done (in a new commit, as the previous version used `<` too).

antlechner committed:
Make precondition on min/max values less strict

There is no reason why this function shouldn't work to "choose" between
just one value.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 63d9f5007744c2812262f0a96b696acec3f40d71).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105439082
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.
